pkg/storage/unified/apistore/storage.go

```diff
@@ -0,0 +1,535 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+// Provenance-includes-location: https://github.com/kubernetes-sigs/apiserver-runtime/blob/main/pkg/experimental/storage/filepath/jsonfile_rest.go
+// Provenance-includes-license: Apache-2.0
+// Provenance-includes-copyright: The Kubernetes Authors.
+
+package apistore
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"reflect"
+	"strconv"
+	"strings"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/conversion"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/selection"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/apiserver/pkg/endpoints/request"
+	"k8s.io/apiserver/pkg/storage"
+	"k8s.io/apiserver/pkg/storage/storagebackend"
+	"k8s.io/apiserver/pkg/storage/storagebackend/factory"
+
+	"github.com/grafana/grafana/pkg/apimachinery/utils"
+	"github.com/grafana/grafana/pkg/storage/unified/resource"
+)
+
+const SortByKey = "grafana.app/sortBy"
+
+var _ storage.Interface = (*Storage)(nil)
+
+// Storage implements storage.Interface and stores resources in unified storage
+type Storage struct {
+	config       *storagebackend.ConfigForResource
+	store        resource.ResourceStoreClient
+	gr           schema.GroupResource
+	codec        runtime.Codec
+	keyFunc      func(obj runtime.Object) (string, error)
+	newFunc      func() runtime.Object
+	newListFunc  func() runtime.Object
+	getAttrsFunc storage.AttrFunc
+	// trigger      storage.IndexerFuncs
+	// indexers     *cache.Indexers
+}
+
+func NewStorage(
+	config *storagebackend.ConfigForResource,
+	gr schema.GroupResource,
+	store resource.ResourceStoreClient,
+	codec runtime.Codec,
+	keyFunc func(obj runtime.Object) (string, error),
+	newFunc func() runtime.Object,
+	newListFunc func() runtime.Object,
+	getAttrsFunc storage.AttrFunc,
+) (storage.Interface, factory.DestroyFunc, error) {
+	return &Storage{
+		config:       config,
+		gr:           gr,
+		codec:        codec,
+		store:        store,
+		keyFunc:      keyFunc,
+		newFunc:      newFunc,
+		newListFunc:  newListFunc,
+		getAttrsFunc: getAttrsFunc,
+	}, nil, nil
+}
+
+func errorWrap(status *resource.StatusResult) error {
+	if status != nil {
+		return &apierrors.StatusError{ErrStatus: metav1.Status{
+			Status:  metav1.StatusFailure,
+			Code:    status.Code,
+			Reason:  metav1.StatusReason(status.Reason),
+			Message: status.Message,
+		}}
+	}
+	return nil
+}
+
+func getKey(ctx context.Context) (*resource.ResourceKey, error) {
+	requestInfo, ok := request.RequestInfoFrom(ctx)
```

Comment:
TODO: We should avoid relying on the key from the request context
https://github.com/grafana/grafana/pull/89635/files

Reply:
done